rfcs/ALL/Access Control Rules.md

```diff
@@ -0,0 +1,134 @@
+* **key_exprs**: supports values of any key type or key-expression (set of keys) type, eg: `temp/room_1`, `temp/**` etc
+* **interfaces**: supports all possible value for network interfaces, eg: `lo`, `lo0` etc
+
+
+For each combination of *interface* + *flow* + *action*  (example:`l0`+`egress`+`put`) in the rules vector, we construct *allow* and *deny* KeTrees (key-expression tries). The *allow* KeTree is built using all the key-expressions provided in the config on which that *interface* is allowed to do that *action* on a particular *flow*. On receiving an authorization request, the key-expression in the request is matched against the appropriate KeTrees to confirm authorization.
+
+## Priority 
+For each message, the access control checks are made in the following order of priority:
+
+*explicit deny* rule > *explicit allow* rule > *default_permission* rule
+
+1. *Explicit deny* rules are given topmost priority. This means that if the key-expression in the  authorization request matches against anything in the *deny* KeTree, it will be denied immediately. If there is no match, then the *explicit allow* rules are checked.
+2. *Explicit allow* permissions come next, i.e., if request matches anything in the *allow* tree then it will be allowed.
+3. The *default_permission* value has least priority. It is applied only if there are no matches in the previous two steps.
+    
+    Note: if *default_permission* is set to `allow`, then there is no need for setting *explicit allow* rules, since they will be discarded anyways.
+    
+
+The decision logic is as follows:
+
+```rust
+fn is_allowed(key_expr) -> decision {
+   if default_permission == deny {
+      if !deny_ketree.match(key_expr) && allow_ketree.match(key_expr) {
+            return true;
+      } else {
+          return false;
+      }
+   } else { //default_permission = allow
+      if deny_ketree.match(key_expr) {
+         return false;
+      } 
+			//no need to check for allow_ketree here
+			return true; 
+   }
+}
+```
+
+## Key-Expression Matching
+
+An important thing to note here is how our key-expression matching works since it ultimately decides the behavior of the access control logic. In matching a key-expression against a KeTree, it will match as a positive only if it is *included in* (*equal to* or *subset of*) the key-expressions specified in the KeTree. A partial match or being a superset will not result in a match.
+
+The following table demonstrates how the matching will work on a key-expression(KE) in request and in the rules set:
+
+| KE in request | KE in rules set | Match |
+|---------------|---------------|-------|
+| test/demo/a         | test/demo/a         | yes   |
+| test/demo/a         | test/demo/*         | yes   |
+| test/\*/\*         | test/**          | yes   |
+| test/demo/a         | **            | yes   |
+| test/demo/a         | test/*/a         | yes   |
+| test/demo/*         | test/*/a         | no    |
```

Comment:
Maybe add "(partial match)" next to the `no`to justify why it does not work.
Same for the following line, can add "(superset)"
Same for the other examples of this RFC.